In [1193]:
import pandas as pd
import mysql.connector
import os
import numpy as np
import warnings
from sklearn.metrics import classification_report

warnings.filterwarnings('ignore')


# Datos sucios

In [1194]:
patients_batch1 = pd.read_csv("../data/patients_batch1.csv")
patients_batch1

ehr  birth_date diagnosis_date death_date
0     10011773  1959-07-05     2015-04-02        NaN
1     10020495  1953-10-02     2017-12-04        NaN
2     10030299  1966-08-16     2019-06-27        NaN
3     10030824  1953-03-03     2018-09-07        NaN
4     10041592  1959-07-06     2018-11-19        NaN
...        ...         ...            ...        ...
7495  77575810  1958-03-21     2014-02-22        NaN
7496  77578551  1948-05-18     2018-07-11        NaN
7497  77579212  1976-02-13     2022-03-22        NaN
7498  77605742  1950-03-16     2015-06-29        NaN
7499  77606485  1962-05-13     2013-11-26        NaN

[7500 rows x 4 columns]

In [1195]:
patients_batch2 = pd.read_csv("../data/patients_batch2.csv")
patients_batch2

ehr  birth_date diagnosis_date  death_date
0     72992494  13/02/1942     08/05/2018         NaN
1     73002338  22/10/1949     08/09/2018         NaN
2     73008149  04/12/1969     12/04/2018         NaN
3     73012939  18/09/1973     17/04/2014  01/01/2018
4     73019870  07/12/1955     24/05/2018         NaN
...        ...         ...            ...         ...
2995  99966287  16/12/1978     02/08/2018         NaN
2996  99981192  10/04/1949     28/10/2017         NaN
2997  99988958  21/01/1966     27/05/2013         NaN
2998  99992350  04/09/1939     17/12/2018         NaN
2999  99994417  10/02/1962     01/04/2017         NaN

[3000 rows x 4 columns]

In [1196]:
gynecological = pd.read_excel("../data/gynecological.xlsx")
gynecological

Unnamed: 0       ehr  pregnancy  birth  caesarean  abort  menarche_age  \
0              0  10011773       -8.0    0.0        NaN    0.0           NaN   
1              1  10030299        3.0    NaN        0.0    0.0           NaN   
2              2  10030824        0.0    0.0        0.0    NaN           NaN   
3              3  10053435        2.0    NaN        0.0    1.0          21.0   
4              4  10111454        0.0   -6.0        NaN    0.0          16.0   
...          ...       ...        ...    ...        ...    ...           ...   
7681        7681  99948591        9.0    2.0       -4.0    0.0          12.0   
7682        7682  99961100        2.0    2.0        0.0    0.0          19.0   
7683        7683  99981192        3.0    3.0        0.0   -5.0          11.0   
7684        7684  99992350        4.0    3.0        NaN   -6.0          11.0   
7685        7685  99994417        2.0   -6.0        NaN    NaN          13.0   

      menopause_age  
0              71.0  
1               NaN  
2              44.0  
3              74.0  
4              55.0  
...             ...  
7681           47.0  
7682           52.0  
7683            NaN  
7684           55.0  
7685            NaN  

[7686 rows x 8 columns]

In [1197]:
histochemistry = pd.read_json("../data/histochemistry.json")
histochemistry

ehr   er  her2  ki67   pr
0     10011773  0.0   0.0  19.0  NaN
1     10020495  1.0   0.0   9.0  0.0
2     10030299  1.0   NaN  18.0  1.0
3     10030824  1.0   0.0   NaN  1.0
4     10041592  0.0   0.0  65.0  0.0
...        ...  ...   ...   ...  ...
9995  99966287  1.0   0.0   NaN  0.0
9996  99981192  1.0   0.0  14.0  0.0
9997  99988958  NaN   0.0  67.0  NaN
9998  99992350  0.0   0.0  16.0  0.0
9999  99994417  1.0   0.0  20.0  1.0

[10000 rows x 5 columns]

In [1198]:
if os.path.exists('../data/tumor.xlsx'):
    tumor = pd.read_excel("../data/tumor.xlsx")
else:
    config = {'host': "localhost",
              'port': 3306,
              'user': "cancer_mama_user",
              'passwd': "cancer_mama_password",
              'db': "cancer_mama"}
    conn = mysql.connector.connect(**config)
    cursor = conn.cursor()

    sql_columns_names = "SHOW COLUMNS from tumor;"
    sql_tumor = "SELECT * FROM tumor"

    cursor.execute(sql_tumor)
    tumor = pd.DataFrame(cursor.fetchall())

    cursor.execute(sql_columns_names)
    columns_names = pd.DataFrame(cursor.fetchall())

    cursor.close()
    conn.close()

    tumor = tumor.set_axis(columns_names[0], axis=1)
    tumor.to_excel("../data/tumor.xlsx")
tumor

Unnamed: 0       ehr  n_tumor t_category  n_category  m_category  \
0               0  10011773        1         IS           0           0   
1               1  10020495        1          1           0           0   
2               2  10020495        2          3           1           0   
3               3  10030299        1          1           0           0   
4               4  10030824        1          2           1           0   
...           ...       ...      ...        ...         ...         ...   
11162       11162  99966287        1          1           1           0   
11163       11163  99981192        1          1           0           0   
11164       11164  99988958        1          0           2           0   
11165       11165  99992350        1          2           0           0   
11166       11166  99994417        1          0           0           0   

      t_category_after_neoadj  n_category_after_neoadj  \
0                         NaN                      NaN   
1                         NaN                      NaN   
2                           2                      0.0   
3                         NaN                      NaN   
4                           2                      2.0   
...                       ...                      ...   
11162                       1                      0.0   
11163                     NaN                      NaN   
11164                       1                      1.0   
11165                       3                      2.0   
11166                       1                      0.0   

       m_category_after_neoadj stage_diagnosis stage_after_neo  grade  ductal  \
0                          NaN               0             NaN      1     NaN   
1                          NaN              IA             NaN      2     1.0   
2                          NaN              IA             NaN      2     1.0   
3                          NaN              IA             NaN      1     1.0   
4                          0.0            IIIA            IIIA      3     NaN   
...                        ...             ...             ...    ...     ...   
11162                      0.0              IB             IIA      3     1.0   
11163                      NaN              IA             NaN      2     NaN   
11164                      0.0            IIIA             IIA      1     1.0   
11165                      0.0             IIA            IIIA      2     1.0   
11166                      0.0              IA              IA      3     NaN   

       lobular neoadjuvant  
0          NaN          no  
1          NaN          no  
2          NaN         yes  
3          NaN          no  
4          1.0         yes  
...        ...         ...  
11162      NaN         yes  
11163      NaN          no  
11164      NaN         yes  
11165      NaN         yes  
11166      NaN         yes  

[11167 rows x 15 columns]

# Preprocesamiento

## Patients

In [1199]:
len(patients_batch1.ehr.unique())

7500

In [1200]:
len(patients_batch2.ehr.unique())

3000

Vemos si hay algún paciente común en ambos batchs de datos de pacientes

In [1201]:
any(patients_batch1['ehr'].isin(patients_batch2['ehr']))

True

Obtenemos los ids de los pacientes que aparecen en ambos 

In [1202]:
elementos_comunes = set(patients_batch1['ehr']).intersection(set(patients_batch2['ehr']))
print(f"El número total de pacientes comunes son: {len(elementos_comunes)}")

El número total de pacientes comunes son: 500


Vemos ahora si estos pacientes que aparecen en ambos documentos tienen valores distintos en cada uno de los documentos

In [1203]:
pacientes_comunes_batch1 = patients_batch1[patients_batch1['ehr'].isin(elementos_comunes)]
pacientes_comunes_batch1

ehr  birth_date diagnosis_date  death_date
7000  72992494  1942-02-13     2018-05-08         NaN
7001  73002338  1949-10-22     2018-09-08         NaN
7002  73008149  1969-12-04     2018-04-12         NaN
7003  73012939  1973-09-18     2014-04-17  2018-01-01
7004  73019870  1955-12-07     2018-05-24         NaN
...        ...         ...            ...         ...
7495  77575810  1958-03-21     2014-02-22         NaN
7496  77578551  1948-05-18     2018-07-11         NaN
7497  77579212  1976-02-13     2022-03-22         NaN
7498  77605742  1950-03-16     2015-06-29         NaN
7499  77606485  1962-05-13     2013-11-26         NaN

[500 rows x 4 columns]

In [1204]:
pacientes_comunes_batch2 = patients_batch2[patients_batch2['ehr'].isin(elementos_comunes)]
pacientes_comunes_batch2

ehr  birth_date diagnosis_date  death_date
0    72992494  13/02/1942     08/05/2018         NaN
1    73002338  22/10/1949     08/09/2018         NaN
2    73008149  04/12/1969     12/04/2018         NaN
3    73012939  18/09/1973     17/04/2014  01/01/2018
4    73019870  07/12/1955     24/05/2018         NaN
..        ...         ...            ...         ...
495  77575810  21/03/1958     22/02/2014         NaN
496  77578551  18/05/1948     11/07/2018         NaN
497  77579212  13/02/1976     22/03/2022         NaN
498  77605742  16/03/1950     29/06/2015         NaN
499  77606485  13/05/1962     26/11/2013         NaN

[500 rows x 4 columns]

Parece ser que los valores sí que son iguales, pero cada documento usa un formato distinto para las fechas. Vamos a ponerlo todo en el formato del primer documento, que es el formato por defecto de datetime: aaaa-mm-dd

In [1205]:
pacientes_comunes_batch2['birth_date'] = pd.to_datetime(pacientes_comunes_batch2['birth_date'], format='%d/%m/%Y')
pacientes_comunes_batch2['death_date'] = pd.to_datetime(pacientes_comunes_batch2['death_date'], format='%d/%m/%Y')
pacientes_comunes_batch2['diagnosis_date'] = pd.to_datetime(pacientes_comunes_batch2['diagnosis_date'], format='%d/%m/%Y')

pacientes_comunes_batch1['birth_date'] = pd.to_datetime(pacientes_comunes_batch1['birth_date'])
pacientes_comunes_batch1['diagnosis_date'] = pd.to_datetime(pacientes_comunes_batch1['diagnosis_date'])
pacientes_comunes_batch1['death_date'] = pd.to_datetime(pacientes_comunes_batch1['death_date'])

pacientes_comunes_batch2

ehr birth_date diagnosis_date death_date
0    72992494 1942-02-13     2018-05-08        NaT
1    73002338 1949-10-22     2018-09-08        NaT
2    73008149 1969-12-04     2018-04-12        NaT
3    73012939 1973-09-18     2014-04-17 2018-01-01
4    73019870 1955-12-07     2018-05-24        NaT
..        ...        ...            ...        ...
495  77575810 1958-03-21     2014-02-22        NaT
496  77578551 1948-05-18     2018-07-11        NaT
497  77579212 1976-02-13     2022-03-22        NaT
498  77605742 1950-03-16     2015-06-29        NaT
499  77606485 1962-05-13     2013-11-26        NaT

[500 rows x 4 columns]

In [1206]:
pacientes_comunes_batch1['birth_date'] = pd.to_datetime(pacientes_comunes_batch1['birth_date'])
pacientes_comunes_batch1['diagnosis_date'] = pd.to_datetime(pacientes_comunes_batch1['diagnosis_date'])
pacientes_comunes_batch1['death_date'] = pd.to_datetime(pacientes_comunes_batch1['death_date'])

In [1207]:
pacientes_comunes_batch1.reset_index(drop=True, inplace=True)

In [1208]:
pacientes_comunes_batch2.equals(pacientes_comunes_batch1)

True

Vemos que todos los valores de los pacientes que aparecen en los dos documentos son exactamente iguales

In [1209]:
patients_batch1

ehr  birth_date diagnosis_date death_date
0     10011773  1959-07-05     2015-04-02        NaN
1     10020495  1953-10-02     2017-12-04        NaN
2     10030299  1966-08-16     2019-06-27        NaN
3     10030824  1953-03-03     2018-09-07        NaN
4     10041592  1959-07-06     2018-11-19        NaN
...        ...         ...            ...        ...
7495  77575810  1958-03-21     2014-02-22        NaN
7496  77578551  1948-05-18     2018-07-11        NaN
7497  77579212  1976-02-13     2022-03-22        NaN
7498  77605742  1950-03-16     2015-06-29        NaN
7499  77606485  1962-05-13     2013-11-26        NaN

[7500 rows x 4 columns]

In [1210]:
patients_batch2

ehr  birth_date diagnosis_date  death_date
0     72992494  13/02/1942     08/05/2018         NaN
1     73002338  22/10/1949     08/09/2018         NaN
2     73008149  04/12/1969     12/04/2018         NaN
3     73012939  18/09/1973     17/04/2014  01/01/2018
4     73019870  07/12/1955     24/05/2018         NaN
...        ...         ...            ...         ...
2995  99966287  16/12/1978     02/08/2018         NaN
2996  99981192  10/04/1949     28/10/2017         NaN
2997  99988958  21/01/1966     27/05/2013         NaN
2998  99992350  04/09/1939     17/12/2018         NaN
2999  99994417  10/02/1962     01/04/2017         NaN

[3000 rows x 4 columns]

In [1211]:
patients_batch2['birth_date'] = pd.to_datetime(patients_batch2['birth_date'], format='%d/%m/%Y')
patients_batch2['diagnosis_date'] = pd.to_datetime(patients_batch2['diagnosis_date'], format='%d/%m/%Y')
patients_batch2['birth_date'] = pd.to_datetime(patients_batch2['birth_date'], format='%d/%m/%Y')

patients_batch1['birth_date'] = pd.to_datetime(patients_batch1['birth_date'])
patients_batch1['diagnosis_date'] = pd.to_datetime(patients_batch1['diagnosis_date'])
patients_batch1['death_date'] = pd.to_datetime(patients_batch1['death_date'])

In [1212]:
patients_batch1_sin_comunes = patients_batch1[~patients_batch1["ehr"].isin(pacientes_comunes_batch1["ehr"])]
patients_batch1_sin_comunes

ehr birth_date diagnosis_date death_date
0     10011773 1959-07-05     2015-04-02        NaT
1     10020495 1953-10-02     2017-12-04        NaT
2     10030299 1966-08-16     2019-06-27        NaT
3     10030824 1953-03-03     2018-09-07        NaT
4     10041592 1959-07-06     2018-11-19        NaT
...        ...        ...            ...        ...
6995  72964107 1959-01-01     2019-12-18        NaT
6996  72975846 1959-07-02     2017-09-01        NaT
6997  72976171 1972-12-01     2018-06-09        NaT
6998  72983987 1962-08-28     2013-07-03 2019-01-25
6999  72990296 1975-11-20     2018-06-17        NaT

[7000 rows x 4 columns]

In [1213]:
patients = pd.concat([patients_batch1_sin_comunes, patients_batch2], ignore_index=True)
patients

ehr birth_date diagnosis_date death_date
0     10011773 1959-07-05     2015-04-02        NaN
1     10020495 1953-10-02     2017-12-04        NaN
2     10030299 1966-08-16     2019-06-27        NaN
3     10030824 1953-03-03     2018-09-07        NaN
4     10041592 1959-07-06     2018-11-19        NaN
...        ...        ...            ...        ...
9995  99966287 1978-12-16     2018-08-02        NaN
9996  99981192 1949-04-10     2017-10-28        NaN
9997  99988958 1966-01-21     2013-05-27        NaN
9998  99992350 1939-09-04     2018-12-17        NaN
9999  99994417 1962-02-10     2017-04-01        NaN

[10000 rows x 4 columns]

In [1214]:
len(patients.ehr.unique())

10000

Ahora deberemos sustituir las columnas de fechas por unas numericas. De esta forma 'birth_date' sera 'age' (edad actual del paciente), 'diagnosis_date' sera 'diagnosis_age' (edad a la que fue diagnsoticado).

In [1215]:
from datetime import datetime
patients['age'] = datetime.now().year - patients['birth_date'].dt.year
patients['diagnosis_age'] = patients['diagnosis_date'].dt.year - patients['birth_date'].dt.year


Por ultimo, hay muchos valores nulos en la variable 'death_date'. Esto sinifica que la mayoria de los pacientes no han muerto. Prodedemos a cambiar esta variable por una binaria que indique tan solo si el paciente ha fallecido o no.

In [1216]:
patients.isnull().sum()

ehr                  0
birth_date           0
diagnosis_date       0
death_date        8865
age                  0
diagnosis_age        0
dtype: int64

In [1217]:
patients['dead'] = patients['death_date'].notnull().astype(int)
patients.drop(columns=['death_date', 'birth_date', 'diagnosis_date'], inplace=True)

In [1218]:
patients

ehr  age  diagnosis_age  dead
0     10011773   65             56     0
1     10020495   71             64     0
2     10030299   58             53     0
3     10030824   71             65     0
4     10041592   65             59     0
...        ...  ...            ...   ...
9995  99966287   46             40     0
9996  99981192   75             68     0
9997  99988958   58             47     0
9998  99992350   85             79     0
9999  99994417   62             55     0

[10000 rows x 4 columns]

## Histochemistry

In [1219]:
histochemistry

ehr   er  her2  ki67   pr
0     10011773  0.0   0.0  19.0  NaN
1     10020495  1.0   0.0   9.0  0.0
2     10030299  1.0   NaN  18.0  1.0
3     10030824  1.0   0.0   NaN  1.0
4     10041592  0.0   0.0  65.0  0.0
...        ...  ...   ...   ...  ...
9995  99966287  1.0   0.0   NaN  0.0
9996  99981192  1.0   0.0  14.0  0.0
9997  99988958  NaN   0.0  67.0  NaN
9998  99992350  0.0   0.0  16.0  0.0
9999  99994417  1.0   0.0  20.0  1.0

[10000 rows x 5 columns]

In [1220]:
histochemistry.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ehr     10000 non-null  int64  
 1   er      9042 non-null   float64
 2   her2    9006 non-null   float64
 3   ki67    9032 non-null   float64
 4   pr      9018 non-null   float64
dtypes: float64(4), int64(1)
memory usage: 390.8 KB


Todos los valores estan dentro de los rangos posibles salvo la variable 'ki67' que al ser un valor porcentual no podria sobrepasar el valor 100, y sin embargo vemos que su valor maximo es 137.

In [1221]:
histochemistry.describe()

ehr           er         her2         ki67           pr
count  1.000000e+04  9042.000000  9006.000000  9032.000000  9018.000000
mean   5.494471e+07     0.662906     0.158894    21.315766     0.665669
std    2.609028e+07     0.472743     0.365598    22.242693     0.471782
min    1.001177e+07     0.000000     0.000000   -57.000000     0.000000
25%    3.214794e+07     0.000000     0.000000    12.000000     0.000000
50%    5.480334e+07     1.000000     0.000000    18.000000     1.000000
75%    7.761323e+07     1.000000     0.000000    28.000000     1.000000
max    9.999442e+07     1.000000     1.000000   137.000000     1.000000

En todas la columnas nos encontramos con valores nulos.

In [1222]:
histochemistry.isnull().sum()

ehr       0
er      958
her2    994
ki67    968
pr      982
dtype: int64

No hay ninguna fila completamente vacia

In [1223]:
filas_nulas = histochemistry[histochemistry.columns.drop('ehr')].isnull().all(axis=1)
filas_nulas[filas_nulas].index.tolist()

[]

Sin embargo hay 488 filas con 2 o mas valores nulos. Procedemos a eliminarlas puesto que este df tiene tan solo 4 variables y si faltan la mitad o mas consideramos que no aporta informacion sufieciente. Ademas tan solo hay 488 filas a las que les sucede, que corresponden tan solo a un 5% del dataset. 

In [1224]:
filas_nulas = histochemistry[histochemistry.columns.drop('ehr')].isnull().sum(axis=1) >=2
len(filas_nulas[filas_nulas].index.tolist())

488

In [1225]:
histochemistry.drop(filas_nulas[filas_nulas].index.tolist(), inplace=True)

Sigue habiendo valores nulos, pero al menos ahora sabemos que en cada fila como mucho hay tan solo 1 valor nulo

In [1226]:
histochemistry.isnull().sum()

ehr       0
er      699
her2    757
ki67    702
pr      725
dtype: int64

In [1227]:
len(histochemistry.ehr.unique())

9512

Coinciden todos los ids con los de patients

In [1228]:
elementos_comunes = set(histochemistry['ehr']).intersection(set(patients['ehr']))
len(elementos_comunes)

9512

Como comparten todos los ids simplemente añadimos la informacion del nuevo df al de 'patients'

In [1229]:
patients = pd.merge(patients, histochemistry, on='ehr')

In [1230]:
patients

ehr  age  diagnosis_age  dead   er  her2  ki67   pr
0     10011773   65             56     0  0.0   0.0  19.0  NaN
1     10020495   71             64     0  1.0   0.0   9.0  0.0
2     10030299   58             53     0  1.0   NaN  18.0  1.0
3     10030824   71             65     0  1.0   0.0   NaN  1.0
4     10041592   65             59     0  0.0   0.0  65.0  0.0
...        ...  ...            ...   ...  ...   ...   ...  ...
9507  99963879   65             62     0  1.0   0.0  18.0  1.0
9508  99966287   46             40     0  1.0   0.0   NaN  0.0
9509  99981192   75             68     0  1.0   0.0  14.0  0.0
9510  99992350   85             79     0  0.0   0.0  16.0  0.0
9511  99994417   62             55     0  1.0   0.0  20.0  1.0

[9512 rows x 8 columns]

## Gynecological

In [1231]:
gynecological

Unnamed: 0       ehr  pregnancy  birth  caesarean  abort  menarche_age  \
0              0  10011773       -8.0    0.0        NaN    0.0           NaN   
1              1  10030299        3.0    NaN        0.0    0.0           NaN   
2              2  10030824        0.0    0.0        0.0    NaN           NaN   
3              3  10053435        2.0    NaN        0.0    1.0          21.0   
4              4  10111454        0.0   -6.0        NaN    0.0          16.0   
...          ...       ...        ...    ...        ...    ...           ...   
7681        7681  99948591        9.0    2.0       -4.0    0.0          12.0   
7682        7682  99961100        2.0    2.0        0.0    0.0          19.0   
7683        7683  99981192        3.0    3.0        0.0   -5.0          11.0   
7684        7684  99992350        4.0    3.0        NaN   -6.0          11.0   
7685        7685  99994417        2.0   -6.0        NaN    NaN          13.0   

      menopause_age  
0              71.0  
1               NaN  
2              44.0  
3              74.0  
4              55.0  
...             ...  
7681           47.0  
7682           52.0  
7683            NaN  
7684           55.0  
7685            NaN  

[7686 rows x 8 columns]

In [1232]:
len(gynecological.ehr.unique())


7686

In [1233]:
gynecological.drop(columns=['Unnamed: 0'], inplace=True)

In [1234]:
gynecological.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7686 entries, 0 to 7685
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   ehr            7686 non-null   int64  
 1   pregnancy      6413 non-null   float64
 2   birth          6417 non-null   float64
 3   caesarean      6131 non-null   float64
 4   abort          6235 non-null   float64
 5   menarche_age   4901 non-null   float64
 6   menopause_age  4486 non-null   float64
dtypes: float64(6), int64(1)
memory usage: 420.5 KB


Observamos como hay valores fuera de rango. Para las variables 'pregnancy', 'birth', 'caesarean', 'abort' los valores minimos son negativos, y los maximos dudamos que puedan ser reales. Para la variable 'menopause_age' el valor maximo es 145, lo cual es imposible. Habra que tratar estos valores asi como deberemos asegurarnos que el valor de pregnancy = birth + caesarean + abort.

In [1235]:
gynecological.describe()

ehr    pregnancy        birth    caesarean        abort  \
count  7.686000e+03  6413.000000  6417.000000  6131.000000  6235.000000   
mean   5.505640e+07     2.078590     1.673991     0.029848     0.259984   
std    2.621591e+07     3.052473     3.006228     2.677790     2.691482   
min    1.001177e+07   -10.000000   -10.000000   -11.000000   -12.000000   
25%    3.202165e+07     1.000000     0.000000     0.000000     0.000000   
50%    5.503108e+07     2.000000     2.000000     0.000000     0.000000   
75%    7.794185e+07     3.000000     3.000000     0.000000     1.000000   
max    9.999442e+07    14.000000    15.000000    12.000000    12.000000   

       menarche_age  menopause_age  
count   4901.000000    4486.000000  
mean      12.736584      50.263041  
std        3.062710      17.196300  
min        0.000000       2.000000  
25%       11.000000      44.000000  
50%       13.000000      49.000000  
75%       14.000000      53.000000  
max       26.000000     145.000000

Vamos a estudiar qué hacer con los valores fuera de rango. Comenzamos con la clase 'pregnancy', la que tenemos 585 filas en las que hay un numero negativo de embarazos, lo cual es imposible. De momento estableceremos estos valores como nulos para su posterior tratamiento. 

In [1236]:
print('Numero de pacientes con numero de embarazos fuera de rango: ', len(gynecological[gynecological['pregnancy'] < 0] ))
gynecological.loc[gynecological['pregnancy'] < 0, 'pregnancy'] = pd.NA

Numero de pacientes con numero de embarazos fuera de rango:  585


Analogamente para las variables 'birth', 'caesarean', 'abort'

In [1237]:
print('Numero de pacientes con numero de partos fuera de rango: ', len(gynecological[gynecological['birth'] < 0] ))
gynecological.loc[gynecological['birth'] < 0, 'birth'] = pd.NA

print('Numero de pacientes con numero de cesareas fuera de rango: ', len(gynecological[gynecological['caesarean'] < 0] ))
gynecological.loc[gynecological['caesarean'] < 0, 'caesarean'] = pd.NA

print('Numero de pacientes con numero de abortos fuera de rango: ', len(gynecological[gynecological['abort'] < 0] ))
gynecological.loc[gynecological['abort'] < 0, 'abort'] = pd.NA


Numero de pacientes con numero de partos fuera de rango:  609
Numero de pacientes con numero de cesareas fuera de rango:  667
Numero de pacientes con numero de abortos fuera de rango:  617


Hay una gran cantidad de nulos, sobre todo para la variable 'menopause_age', siendo el 41.6% de los valores nulos. Teniendo esto en cuenta procederemos a no considerar esta variable. 

In [1238]:
gynecological.isnull().sum()

ehr                 0
pregnancy        1858
birth            1878
caesarean        2222
abort            2068
menarche_age     2785
menopause_age    3200
dtype: int64

In [1239]:
gynecological.drop(columns='menopause_age', inplace=True)

Una vez hemos eliminado valores fuera de rango consideraremos los valores, que a pesar de estar dentro dee rango, son atipicos

# FALTA IMPUTAR LOS EMBARAZOS Y DESPUES VER Q HACER CON OUTLIERS

LO DE LOS OUTLIERS EL CODIGO ESTA, SOLO FALTA EL DE IMPUTACION


In [1240]:
Q1 = gynecological['pregnancy'].quantile(0.25)
Q3 = gynecological['pregnancy'].quantile(0.75)
IQR = Q3 - Q1

# Definir los límites para considerar un dato como atípico
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Identificar los valores atípicos
outliers = gynecological[(gynecological['pregnancy'] < lower_bound) | (gynecological['pregnancy'] > upper_bound)]
# gynecological = gynecological.drop(outliers.index, inplace=True)


Sin contar la variable identificadora 'ehr' tenemos tan solo 5 variables explicativas en este dataset. Porcedemos a eliminar aquellas filas que tengan tan solo 1 valor no nulo. Podriamos optar a eliminar aquellas con tan solo 2 valores no nulos, pero primero debremos estudiar si podemos imputar nosotros esos valores.

In [1241]:
filas_nulas = gynecological[gynecological.columns.drop('ehr')].isnull().sum(axis=1) >=4
len(filas_nulas[filas_nulas].index.tolist())

135

In [1242]:
gynecological.drop(filas_nulas[filas_nulas].index.tolist(), inplace=True)

In [1243]:
len(gynecological.ehr.unique())


7551

In [1244]:
elementos_comunes = set(gynecological['ehr']).intersection(set(patients['ehr']))
len(elementos_comunes)

7185

In [1245]:
gynecological_patients = pd.merge(gynecological, patients, on='ehr')

In [1246]:
gynecological_patients

ehr  pregnancy  birth  caesarean  abort  menarche_age  age  \
0     10011773        NaN    0.0        NaN    0.0           NaN   65   
1     10030299        3.0    NaN        0.0    0.0           NaN   58   
2     10030824        0.0    0.0        0.0    NaN           NaN   71   
3     10053435        2.0    NaN        0.0    1.0          21.0   67   
4     10111454        0.0    NaN        NaN    0.0          16.0   63   
...        ...        ...    ...        ...    ...           ...  ...   
7180  99948591        9.0    2.0        NaN    0.0          12.0   69   
7181  99961100        2.0    2.0        0.0    0.0          19.0   54   
7182  99981192        3.0    3.0        0.0    NaN          11.0   75   
7183  99992350        4.0    3.0        NaN    NaN          11.0   85   
7184  99994417        2.0    NaN        NaN    NaN          13.0   62   

      diagnosis_age  dead   er  her2  ki67   pr  
0                56     0  0.0   0.0  19.0  NaN  
1                53     0  1.0   NaN  18.0  1.0  
2                65     0  1.0   0.0   NaN  1.0  
3                60     0  1.0   0.0   NaN  1.0  
4                56     0  1.0   0.0  72.0  1.0  
...             ...   ...  ...   ...   ...  ...  
7180             67     0  1.0   0.0  12.0  1.0  
7181             48     0  1.0   0.0  21.0  1.0  
7182             68     0  1.0   0.0  14.0  0.0  
7183             79     0  0.0   0.0  16.0  0.0  
7184             55     0  1.0   0.0  20.0  1.0  

[7185 rows x 13 columns]

## Tumor

In [1247]:
tumor

Unnamed: 0       ehr  n_tumor t_category  n_category  m_category  \
0               0  10011773        1         IS           0           0   
1               1  10020495        1          1           0           0   
2               2  10020495        2          3           1           0   
3               3  10030299        1          1           0           0   
4               4  10030824        1          2           1           0   
...           ...       ...      ...        ...         ...         ...   
11162       11162  99966287        1          1           1           0   
11163       11163  99981192        1          1           0           0   
11164       11164  99988958        1          0           2           0   
11165       11165  99992350        1          2           0           0   
11166       11166  99994417        1          0           0           0   

      t_category_after_neoadj  n_category_after_neoadj  \
0                         NaN                      NaN   
1                         NaN                      NaN   
2                           2                      0.0   
3                         NaN                      NaN   
4                           2                      2.0   
...                       ...                      ...   
11162                       1                      0.0   
11163                     NaN                      NaN   
11164                       1                      1.0   
11165                       3                      2.0   
11166                       1                      0.0   

       m_category_after_neoadj stage_diagnosis stage_after_neo  grade  ductal  \
0                          NaN               0             NaN      1     NaN   
1                          NaN              IA             NaN      2     1.0   
2                          NaN              IA             NaN      2     1.0   
3                          NaN              IA             NaN      1     1.0   
4                          0.0            IIIA            IIIA      3     NaN   
...                        ...             ...             ...    ...     ...   
11162                      0.0              IB             IIA      3     1.0   
11163                      NaN              IA             NaN      2     NaN   
11164                      0.0            IIIA             IIA      1     1.0   
11165                      0.0             IIA            IIIA      2     1.0   
11166                      0.0              IA              IA      3     NaN   

       lobular neoadjuvant  
0          NaN          no  
1          NaN          no  
2          NaN         yes  
3          NaN          no  
4          1.0         yes  
...        ...         ...  
11162      NaN         yes  
11163      NaN          no  
11164      NaN         yes  
11165      NaN         yes  
11166      NaN         yes  

[11167 rows x 15 columns]

In [1248]:
tumor.drop(columns=['Unnamed: 0'], inplace=True)

hay ids que se repiten. una persona puede tener varios tumores

In [1249]:
len(tumor.ehr.unique())

9971

In [1250]:
elementos_comunes = set(tumor['ehr']).intersection(set(gynecological_patients['ehr']))
len(elementos_comunes)

7167

Tiene mas elementos en comun con 'patients' que con 'gynecological_patients'

In [1251]:
elementos_comunes = set(tumor['ehr']).intersection(set(patients['ehr']))
len(elementos_comunes)

9485

In [1252]:
tumor_patients = pd.merge(tumor, patients, on='ehr')

Finalmente obtenemos un df en el que se incluye la informacion de todos los pasados salvo 'gynecological'. Sin embargo en este df no hay una unica fila por cada paciente, sino que puede haber varias debido a que un paciente puede tener varios tumores, dando lugar a una duplicidad parcial de filas

In [1253]:
tumor_patients

ehr  n_tumor t_category  n_category  m_category  \
0      10011773        1         IS           0           0   
1      10020495        1          1           0           0   
2      10020495        2          3           1           0   
3      10030299        1          1           0           0   
4      10030824        1          2           1           0   
...         ...      ...        ...         ...         ...   
10621  99963879        1          2           3           0   
10622  99966287        1          1           1           0   
10623  99981192        1          1           0           0   
10624  99992350        1          2           0           0   
10625  99994417        1          0           0           0   

      t_category_after_neoadj  n_category_after_neoadj  \
0                         NaN                      NaN   
1                         NaN                      NaN   
2                           2                      0.0   
3                         NaN                      NaN   
4                           2                      2.0   
...                       ...                      ...   
10621                       3                      1.0   
10622                       1                      0.0   
10623                     NaN                      NaN   
10624                       3                      2.0   
10625                       1                      0.0   

       m_category_after_neoadj stage_diagnosis stage_after_neo  ...  ductal  \
0                          NaN               0             NaN  ...     NaN   
1                          NaN              IA             NaN  ...     1.0   
2                          NaN              IA             NaN  ...     1.0   
3                          NaN              IA             NaN  ...     1.0   
4                          0.0            IIIA            IIIA  ...     NaN   
...                        ...             ...             ...  ...     ...   
10621                      0.0            IIIC               0  ...     1.0   
10622                      0.0              IB             IIA  ...     1.0   
10623                      NaN              IA             NaN  ...     NaN   
10624                      0.0             IIA            IIIA  ...     1.0   
10625                      0.0              IA              IA  ...     NaN   

       lobular  neoadjuvant age  diagnosis_age  dead   er  her2  ki67   pr  
0          NaN           no  65             56     0  0.0   0.0  19.0  NaN  
1          NaN           no  71             64     0  1.0   0.0   9.0  0.0  
2          NaN          yes  71             64     0  1.0   0.0   9.0  0.0  
3          NaN           no  58             53     0  1.0   NaN  18.0  1.0  
4          1.0          yes  71             65     0  1.0   0.0   NaN  1.0  
...        ...          ...  ..            ...   ...  ...   ...   ...  ...  
10621      NaN          yes  65             62     0  1.0   0.0  18.0  1.0  
10622      NaN          yes  46             40     0  1.0   0.0   NaN  0.0  
10623      NaN           no  75             68     0  1.0   0.0  14.0  0.0  
10624      NaN          yes  85             79     0  0.0   0.0  16.0  0.0  
10625      NaN          yes  62             55     0  1.0   0.0  20.0  1.0  

[10626 rows x 21 columns]

In [1254]:
len(tumor_patients.ehr.unique())

9485

Creamos tambien un df mas reducido pero que sí disponga de los datos de 'gynecological'

In [1255]:
tumor_patients_gyn = pd.merge(tumor, gynecological_patients, on='ehr')

In [1256]:
tumor_patients_gyn

ehr  n_tumor t_category  n_category  m_category  \
0     10011773        1         IS           0           0   
1     10030299        1          1           0           0   
2     10030824        1          2           1           0   
3     10053435        1          0           0           0   
4     10111454        1          1           0           0   
...        ...      ...        ...         ...         ...   
8048  99948591        1          2           0           0   
8049  99961100        1          1           2           0   
8050  99981192        1          1           0           0   
8051  99992350        1          2           0           0   
8052  99994417        1          0           0           0   

     t_category_after_neoadj  n_category_after_neoadj  \
0                        NaN                      NaN   
1                        NaN                      NaN   
2                          2                      2.0   
3                        NaN                      NaN   
4                          1                      0.0   
...                      ...                      ...   
8048                     NaN                      NaN   
8049                       1                      0.0   
8050                     NaN                      NaN   
8051                       3                      2.0   
8052                       1                      0.0   

      m_category_after_neoadj stage_diagnosis stage_after_neo  ...  caesarean  \
0                         NaN               0             NaN  ...        NaN   
1                         NaN              IA             NaN  ...        0.0   
2                         0.0            IIIA            IIIA  ...        0.0   
3                         NaN              IV             NaN  ...        0.0   
4                         0.0              IA              IA  ...        NaN   
...                       ...             ...             ...  ...        ...   
8048                      NaN             IIA             NaN  ...        NaN   
8049                      0.0             IIA              IB  ...        0.0   
8050                      NaN              IA             NaN  ...        0.0   
8051                      0.0             IIA            IIIA  ...        NaN   
8052                      0.0              IA              IA  ...        NaN   

      abort  menarche_age age  diagnosis_age  dead   er  her2  ki67   pr  
0       0.0           NaN  65             56     0  0.0   0.0  19.0  NaN  
1       0.0           NaN  58             53     0  1.0   NaN  18.0  1.0  
2       NaN           NaN  71             65     0  1.0   0.0   NaN  1.0  
3       1.0          21.0  67             60     0  1.0   0.0   NaN  1.0  
4       0.0          16.0  63             56     0  1.0   0.0  72.0  1.0  
...     ...           ...  ..            ...   ...  ...   ...   ...  ...  
8048    0.0          12.0  69             67     0  1.0   0.0  12.0  1.0  
8049    0.0          19.0  54             48     0  1.0   0.0  21.0  1.0  
8050    NaN          11.0  75             68     0  1.0   0.0  14.0  0.0  
8051    NaN          11.0  85             79     0  0.0   0.0  16.0  0.0  
8052    NaN          13.0  62             55     0  1.0   0.0  20.0  1.0  

[8053 rows x 26 columns]

In [1257]:
porcentaje_0 = (tumor_patients_gyn['n_tumor'] == 1).mean() * 100
porcentaje_1 = (tumor_patients_gyn['n_tumor'] == 2).mean() * 100

# Imprime los resultados
print(f'Porcentaje de m_category igual a 0: {porcentaje_0:.2f}%')
print(f'Porcentaje de m_category igual a 1: {porcentaje_1:.2f}%')

Porcentaje de m_category igual a 0: 89.00%
Porcentaje de m_category igual a 1: 11.00%


Nuestro objetivo consiste en predecir si en un paciente habra o no una recaida. No contamos con una variable directamente que nos indique si un paciente recayó o no, pero haciendo uso de este dataset podemos crear una variable asi. 

Un paciente ha recaido si ha tenido mas de un tumor, por lo que podemos construir un nuevo dataframe con una columna binaria 'recaida' si existen 2 filas con mismo 'ehr', ya que cada una correspondera a un 'n_tumor' distinto. Para predecir si un paciente recaera o no tan solo necesitamos la informacion del segundo tumor, por tanto eliminaremos las filas que hacen referencia a los tumores reincidentes y de esta forma habra una unica fila por paciente para poder expresar si el paciente ha recaido o no.

In [1258]:
tumor_patients_recaida = tumor_patients.copy()

# Crear la nueva variable 'recaida'
tumor_patients_recaida['recaida'] = tumor_patients_recaida['n_tumor'].apply(lambda x: 1 if x == 2 else 0)

# Asignar 'recaida' a 1 incluso en el primer valor del paciente
tumor_patients_recaida.loc[tumor_patients_recaida.duplicated(subset=['ehr'], keep='first') | ~tumor_patients_recaida.duplicated(subset=['ehr'], keep=False), 'recaida'] = 1

# Eliminar las filas duplicadas, conservando la primera aparición y la nueva variable 'recaida'
tumor_patients_recaida = tumor_patients_recaida.drop_duplicates(subset=['ehr'], keep='first')

tumor_patients_recaida.drop(columns=['n_tumor'], inplace=True)


In [1259]:
porcentaje_0 = (tumor_patients_recaida['recaida'] == 0).mean() * 100
porcentaje_1 = (tumor_patients_recaida['recaida'] == 1).mean() * 100

# Imprime los resultados
print(f'Porcentaje de pacientes con recaida: {porcentaje_0:.2f}%')
print(f'Porcentaje de pacientes sin recaida: {porcentaje_1:.2f}%')

Porcentaje de pacientes con recaida: 12.03%
Porcentaje de pacientes sin recaida: 87.97%


In [1260]:
tumor_patients_recaida

ehr t_category  n_category  m_category t_category_after_neoadj  \
0      10011773         IS           0           0                     NaN   
1      10020495          1           0           0                     NaN   
3      10030299          1           0           0                     NaN   
4      10030824          2           1           0                       2   
5      10041592          1           1           0                       4   
...         ...        ...         ...         ...                     ...   
10621  99963879          2           3           0                       3   
10622  99966287          1           1           0                       1   
10623  99981192          1           0           0                     NaN   
10624  99992350          2           0           0                       3   
10625  99994417          0           0           0                       1   

       n_category_after_neoadj  m_category_after_neoadj stage_diagnosis  \
0                          NaN                      NaN               0   
1                          NaN                      NaN              IA   
3                          NaN                      NaN              IA   
4                          2.0                      0.0            IIIA   
5                          0.0                      0.0              IB   
...                        ...                      ...             ...   
10621                      1.0                      0.0            IIIC   
10622                      0.0                      0.0              IB   
10623                      NaN                      NaN              IA   
10624                      2.0                      0.0             IIA   
10625                      0.0                      0.0              IA   

      stage_after_neo  grade  ...  lobular  neoadjuvant age  diagnosis_age  \
0                 NaN      1  ...      NaN           no  65             56   
1                 NaN      2  ...      NaN           no  71             64   
3                 NaN      1  ...      NaN           no  58             53   
4                IIIA      3  ...      1.0          yes  71             65   
5                  IA      2  ...      NaN          yes  65             59   
...               ...    ...  ...      ...          ...  ..            ...   
10621               0      2  ...      NaN          yes  65             62   
10622             IIA      3  ...      NaN          yes  46             40   
10623             NaN      2  ...      NaN           no  75             68   
10624            IIIA      2  ...      NaN          yes  85             79   
10625              IA      3  ...      NaN          yes  62             55   

       dead   er  her2  ki67   pr  recaida  
0         0  0.0   0.0  19.0  NaN        1  
1         0  1.0   0.0   9.0  0.0        0  
3         0  1.0   NaN  18.0  1.0        1  
4         0  1.0   0.0   NaN  1.0        1  
5         0  0.0   0.0  65.0  0.0        1  
...     ...  ...   ...   ...  ...      ...  
10621     0  1.0   0.0  18.0  1.0        1  
10622     0  1.0   0.0   NaN  0.0        1  
10623     0  1.0   0.0  14.0  0.0        1  
10624     0  0.0   0.0  16.0  0.0        1  
10625     0  1.0   0.0  20.0  1.0        1  

[9485 rows x 21 columns]

In [1261]:
tumor_patients_recaida.describe()


ehr   n_category   m_category  n_category_after_neoadj  \
count  9.485000e+03  9485.000000  9485.000000              5171.000000   
mean   5.481923e+07     0.847127     0.098577                 0.785728   
std    2.604873e+07     0.988671     0.298109                 0.970013   
min    1.001177e+07     0.000000     0.000000                 0.000000   
25%    3.208911e+07     0.000000     0.000000                 0.000000   
50%    5.464934e+07     1.000000     0.000000                 0.000000   
75%    7.739152e+07     1.000000     0.000000                 1.000000   
max    9.999442e+07     3.000000     1.000000                 3.000000   

       m_category_after_neoadj        grade  ductal  lobular          age  \
count              5171.000000  9485.000000  5383.0   1312.0  9485.000000   
mean                  0.099401     1.939062     1.0      1.0    63.946547   
std                   0.299228     0.697254     0.0      0.0    10.096988   
min                   0.000000     1.000000     1.0      1.0    26.000000   
25%                   0.000000     1.000000     1.0      1.0    57.000000   
50%                   0.000000     2.000000     1.0      1.0    64.000000   
75%                   0.000000     2.000000     1.0      1.0    71.000000   
max                   1.000000     3.000000     1.0      1.0   100.000000   

       diagnosis_age         dead           er         her2         ki67  \
count    9485.000000  9485.000000  8786.000000  8729.000000  8786.000000   
mean       57.088666     0.112599     0.664352     0.158667    21.275780   
std        10.157607     0.316118     0.472243     0.365386    22.252437   
min        14.000000     0.000000     0.000000     0.000000   -57.000000   
25%        50.000000     0.000000     0.000000     0.000000    12.000000   
50%        57.000000     0.000000     1.000000     0.000000    18.000000   
75%        64.000000     0.000000     1.000000     0.000000    28.000000   
max        97.000000     1.000000     1.000000     1.000000   137.000000   

                pr      recaida  
count  8761.000000  9485.000000  
mean      0.666705     0.879705  
std       0.471418     0.325324  
min       0.000000     0.000000  
25%       0.000000     1.000000  
50%       1.000000     1.000000  
75%       1.000000     1.000000  
max       1.000000     1.000000

Como podemos ver hay una variable 'lobular' para la cual faltan el 86% de los datos, por lo que en primera instancia optamos por prescindirs de esta variable. Tambien faltan los valores de 't_category_after_neoadj', 'n_category_after_neoadj', 'm_category_after_neoadj' y 'stage_after_neo' para el 45.4% de los datos. Esto es debido a que estas variables solo toman valores en pacientes a los que se les aplico un tratamiento neoadyuvante, por lo que los que no lo recibieron tienen nulo en todas estas columnas. 

Sabiendo esto podemos optar por no tener en cuenta estas variables, o crear una nueva variable que indique si el paciente recibio o no el tratamiento neoadyuvante.

In [1262]:
tumor_patients_recaida.isnull().sum()

ehr                           0
t_category                    0
n_category                    0
m_category                    0
t_category_after_neoadj    4314
n_category_after_neoadj    4314
m_category_after_neoadj    4314
stage_diagnosis               0
stage_after_neo            4314
grade                         0
ductal                     4102
lobular                    8173
neoadjuvant                   0
age                           0
diagnosis_age                 0
dead                          0
er                          699
her2                        756
ki67                        699
pr                          724
recaida                       0
dtype: int64

En un primer momento parecia que faltaban una gran cantidad de datos para la variable binaria 'lobular', sin embargo si nos fijamos en sus valores todo ellos corresponden al valor '1', por lo que parece que los nulos no lo son realmente, sino que corresponderian a la clase '0'

In [1263]:
print(tumor_patients_recaida.lobular.unique())
tumor_patients_recaida['lobular'] = tumor_patients_recaida['lobular'].notnull().astype(int)

[nan  1.]


Sucede lo mismo para la clase 'ductal'

In [1264]:
print(tumor_patients_recaida.ductal.unique())
tumor_patients_recaida['ductal'] = tumor_patients_recaida['ductal'].notnull().astype(int)

[nan  1.]


En cuanto a las variables 't_category_after_neoadj', 'n_category_after_neoadj', 'm_category_after_neoadj' en primer lugar optamos por no considerarlas.

In [1265]:
tumor_patients_recaida[tumor_patients_recaida[['t_category_after_neoadj', 'n_category_after_neoadj', 'm_category_after_neoadj', 'stage_after_neo']].isnull().all(axis=1)]


ehr t_category  n_category  m_category t_category_after_neoadj  \
0      10011773         IS           0           0                     NaN   
1      10020495          1           0           0                     NaN   
3      10030299          1           0           0                     NaN   
6      10053435          0           0           0                     NaN   
12     10115313          1           0           1                     NaN   
...         ...        ...         ...         ...                     ...   
10613  99859482          0           3           0                     NaN   
10617  99915606         IS           0           0                     NaN   
10618  99948515          1           0           0                     NaN   
10619  99948591          2           0           0                     NaN   
10623  99981192          1           0           0                     NaN   

       n_category_after_neoadj  m_category_after_neoadj stage_diagnosis  \
0                          NaN                      NaN               0   
1                          NaN                      NaN              IA   
3                          NaN                      NaN              IA   
6                          NaN                      NaN              IV   
12                         NaN                      NaN              IA   
...                        ...                      ...             ...   
10613                      NaN                      NaN              IV   
10617                      NaN                      NaN             IIB   
10618                      NaN                      NaN              IA   
10619                      NaN                      NaN             IIA   
10623                      NaN                      NaN              IA   

      stage_after_neo  grade  ...  lobular  neoadjuvant age  diagnosis_age  \
0                 NaN      1  ...        0           no  65             56   
1                 NaN      2  ...        0           no  71             64   
3                 NaN      1  ...        0           no  58             53   
6                 NaN      3  ...        0           no  67             60   
12                NaN      3  ...        0           no  67             62   
...               ...    ...  ...      ...          ...  ..            ...   
10613             NaN      2  ...        0           no  60             58   
10617             NaN      2  ...        0           no  63             56   
10618             NaN      3  ...        1           no  70             64   
10619             NaN      2  ...        1           no  69             67   
10623             NaN      2  ...        0           no  75             68   

       dead   er  her2  ki67   pr  recaida  
0         0  0.0   0.0  19.0  NaN        1  
1         0  1.0   0.0   9.0  0.0        0  
3         0  1.0   NaN  18.0  1.0        1  
6         0  1.0   0.0   NaN  1.0        1  
12        0  1.0   0.0  14.0  0.0        1  
...     ...  ...   ...   ...  ...      ...  
10613     0  NaN   0.0  19.0  1.0        1  
10617     0  0.0   0.0  49.0  1.0        1  
10618     0  0.0   0.0  25.0  1.0        1  
10619     0  1.0   0.0  12.0  1.0        1  
10623     0  1.0   0.0  14.0  0.0        1  

[4314 rows x 21 columns]

Procedemos a eliminar las columnas discutidas.

In [1266]:
tumor_patients_recaida.drop(columns=['t_category_after_neoadj', 'n_category_after_neoadj', 'm_category_after_neoadj','stage_after_neo'], inplace=True)


In [1267]:
tumor_patients_recaida

ehr t_category  n_category  m_category stage_diagnosis  grade  \
0      10011773         IS           0           0               0      1   
1      10020495          1           0           0              IA      2   
3      10030299          1           0           0              IA      1   
4      10030824          2           1           0            IIIA      3   
5      10041592          1           1           0              IB      2   
...         ...        ...         ...         ...             ...    ...   
10621  99963879          2           3           0            IIIC      2   
10622  99966287          1           1           0              IB      3   
10623  99981192          1           0           0              IA      2   
10624  99992350          2           0           0             IIA      2   
10625  99994417          0           0           0              IA      3   

       ductal  lobular neoadjuvant  age  diagnosis_age  dead   er  her2  ki67  \
0           0        0          no   65             56     0  0.0   0.0  19.0   
1           1        0          no   71             64     0  1.0   0.0   9.0   
3           1        0          no   58             53     0  1.0   NaN  18.0   
4           0        1         yes   71             65     0  1.0   0.0   NaN   
5           1        0         yes   65             59     0  0.0   0.0  65.0   
...       ...      ...         ...  ...            ...   ...  ...   ...   ...   
10621       1        0         yes   65             62     0  1.0   0.0  18.0   
10622       1        0         yes   46             40     0  1.0   0.0   NaN   
10623       0        0          no   75             68     0  1.0   0.0  14.0   
10624       1        0         yes   85             79     0  0.0   0.0  16.0   
10625       0        0         yes   62             55     0  1.0   0.0  20.0   

        pr  recaida  
0      NaN        1  
1      0.0        0  
3      1.0        1  
4      1.0        1  
5      0.0        1  
...    ...      ...  
10621  1.0        1  
10622  0.0        1  
10623  0.0        1  
10624  0.0        1  
10625  1.0        1  

[9485 rows x 17 columns]

Como podemos observar las variables 't_category', 'stage_diagnosis' y 'neoadjuvant' son categoricas, por lo que aplicaremos OneHotEncoder para que adopten valores numericos. Para 'neoadjuvant' simplemente daremos el valor '0' a 'no' y '1' a 'yes'.

In [1268]:
from sklearn.preprocessing import OneHotEncoder

tumor_patients_recaida['neoadjuvant'] = tumor_patients_recaida['neoadjuvant'].map({'no': 0, 'yes': 1})

encoder = OneHotEncoder(sparse=False)
columns_to_encode = ['t_category', 'stage_diagnosis']

encoded_columns = encoder.fit_transform(tumor_patients_recaida[columns_to_encode])
# encoded_df = pd.DataFrame(encoded_columns, columns=encoder.get_feature_names_out(columns_to_encode))


# t_catOHE = encoder.fit_transform(tumor_patients_recaida[['t_category']])
# t_cols = encoder.get_feature_names_out(['t_category'])
tumor_patients_recaida[encoder.get_feature_names_out(columns_to_encode)] = encoded_columns
tumor_patients_recaida.drop(columns=['t_category', 'stage_diagnosis'], inplace=True)
tumor_patients_recaida

ehr  n_category  m_category  grade  ductal  lobular  neoadjuvant  \
0      10011773           0           0      1       0        0            0   
1      10020495           0           0      2       1        0            0   
3      10030299           0           0      1       1        0            0   
4      10030824           1           0      3       0        1            1   
5      10041592           1           0      2       1        0            1   
...         ...         ...         ...    ...     ...      ...          ...   
10621  99963879           3           0      2       1        0            1   
10622  99966287           1           0      3       1        0            1   
10623  99981192           0           0      2       0        0            0   
10624  99992350           0           0      2       1        0            1   
10625  99994417           0           0      3       0        0            1   

       age  diagnosis_age  dead  ...  t_category_IS  stage_diagnosis_0  \
0       65             56     0  ...            1.0                1.0   
1       71             64     0  ...            0.0                0.0   
3       58             53     0  ...            0.0                0.0   
4       71             65     0  ...            0.0                0.0   
5       65             59     0  ...            0.0                0.0   
...    ...            ...   ...  ...            ...                ...   
10621   65             62     0  ...            0.0                0.0   
10622   46             40     0  ...            0.0                0.0   
10623   75             68     0  ...            0.0                0.0   
10624   85             79     0  ...            0.0                0.0   
10625   62             55     0  ...            0.0                0.0   

       stage_diagnosis_IA  stage_diagnosis_IB  stage_diagnosis_IIA  \
0                     0.0                 0.0                  0.0   
1                     1.0                 0.0                  0.0   
3                     1.0                 0.0                  0.0   
4                     0.0                 0.0                  0.0   
5                     0.0                 1.0                  0.0   
...                   ...                 ...                  ...   
10621                 0.0                 0.0                  0.0   
10622                 0.0                 1.0                  0.0   
10623                 1.0                 0.0                  0.0   
10624                 0.0                 0.0                  1.0   
10625                 1.0                 0.0                  0.0   

       stage_diagnosis_IIB  stage_diagnosis_IIIA  stage_diagnosis_IIIB  \
0                      0.0                   0.0                   0.0   
1                      0.0                   0.0                   0.0   
3                      0.0                   0.0                   0.0   
4                      0.0                   1.0                   0.0   
5                      0.0                   0.0                   0.0   
...                    ...                   ...                   ...   
10621                  0.0                   0.0                   0.0   
10622                  0.0                   0.0                   0.0   
10623                  0.0                   0.0                   0.0   
10624                  0.0                   0.0                   0.0   
10625                  0.0                   0.0                   0.0   

       stage_diagnosis_IIIC  stage_diagnosis_IV  
0                       0.0                 0.0  
1                       0.0                 0.0  
3                       0.0                 0.0  
4                       0.0                 0.0  
5                       0.0                 0.0  
...                     ...                 ...  
10621                   1.0                 0.0  
10622                   0.0     

In [1269]:
tumor_patients_recaida.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9485 entries, 0 to 10625
Data columns (total 30 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   ehr                   9485 non-null   int64  
 1   n_category            9485 non-null   int64  
 2   m_category            9485 non-null   int64  
 3   grade                 9485 non-null   int64  
 4   ductal                9485 non-null   int32  
 5   lobular               9485 non-null   int32  
 6   neoadjuvant           9485 non-null   int64  
 7   age                   9485 non-null   int64  
 8   diagnosis_age         9485 non-null   int64  
 9   dead                  9485 non-null   int32  
 10  er                    8786 non-null   float64
 11  her2                  8729 non-null   float64
 12  ki67                  8786 non-null   float64
 13  pr                    8761 non-null   float64
 14  recaida               9485 non-null   int64  
 15  t_category_0        

In [1270]:
tumor_patients_recaida.isnull().sum()

ehr                       0
n_category                0
m_category                0
grade                     0
ductal                    0
lobular                   0
neoadjuvant               0
age                       0
diagnosis_age             0
dead                      0
er                      699
her2                    756
ki67                    699
pr                      724
recaida                   0
t_category_0              0
t_category_1              0
t_category_2              0
t_category_3              0
t_category_4              0
t_category_IS             0
stage_diagnosis_0         0
stage_diagnosis_IA        0
stage_diagnosis_IB        0
stage_diagnosis_IIA       0
stage_diagnosis_IIB       0
stage_diagnosis_IIIA      0
stage_diagnosis_IIIB      0
stage_diagnosis_IIIC      0
stage_diagnosis_IV        0
dtype: int64

Como podemos comprobar para cada fila, como maximo, hay un unico valor nulo. Ante esto se nos plantean 2 posibilidades:

* Imputar los valores (KNNImputer)
* Prescindir de estas filas (Nos quedariamos con 9485 - 2878 = 6607 datos.)

In [1271]:
filas_nulas = tumor_patients_recaida[tumor_patients_recaida.columns.drop('ehr')].isnull().sum(axis=1) >=2
len(filas_nulas[filas_nulas].index.tolist())

0

En una primera aproximacion optaremos por imputar los valores mediante KNNImputer

In [1272]:
from sklearn.impute import KNNImputer

imputer = KNNImputer(n_neighbors=2)
tumor_patients_recaida =pd.DataFrame(imputer.fit_transform(tumor_patients_recaida), columns=tumor_patients_recaida.columns)

In [1273]:
tumor_patients_recaida

ehr  n_category  m_category  grade  ductal  lobular  neoadjuvant  \
0     10011773.0         0.0         0.0    1.0     0.0      0.0          0.0   
1     10020495.0         0.0         0.0    2.0     1.0      0.0          0.0   
2     10030299.0         0.0         0.0    1.0     1.0      0.0          0.0   
3     10030824.0         1.0         0.0    3.0     0.0      1.0          1.0   
4     10041592.0         1.0         0.0    2.0     1.0      0.0          1.0   
...          ...         ...         ...    ...     ...      ...          ...   
9480  99963879.0         3.0         0.0    2.0     1.0      0.0          1.0   
9481  99966287.0         1.0         0.0    3.0     1.0      0.0          1.0   
9482  99981192.0         0.0         0.0    2.0     0.0      0.0          0.0   
9483  99992350.0         0.0         0.0    2.0     1.0      0.0          1.0   
9484  99994417.0         0.0         0.0    3.0     0.0      0.0          1.0   

       age  diagnosis_age  dead  ...  t_category_IS  stage_diagnosis_0  \
0     65.0           56.0   0.0  ...            1.0                1.0   
1     71.0           64.0   0.0  ...            0.0                0.0   
2     58.0           53.0   0.0  ...            0.0                0.0   
3     71.0           65.0   0.0  ...            0.0                0.0   
4     65.0           59.0   0.0  ...            0.0                0.0   
...    ...            ...   ...  ...            ...                ...   
9480  65.0           62.0   0.0  ...            0.0                0.0   
9481  46.0           40.0   0.0  ...            0.0                0.0   
9482  75.0           68.0   0.0  ...            0.0                0.0   
9483  85.0           79.0   0.0  ...            0.0                0.0   
9484  62.0           55.0   0.0  ...            0.0                0.0   

      stage_diagnosis_IA  stage_diagnosis_IB  stage_diagnosis_IIA  \
0                    0.0                 0.0                  0.0   
1                    1.0                 0.0                  0.0   
2                    1.0                 0.0                  0.0   
3                    0.0                 0.0                  0.0   
4                    0.0                 1.0                  0.0   
...                  ...                 ...                  ...   
9480                 0.0                 0.0                  0.0   
9481                 0.0                 1.0                  0.0   
9482                 1.0                 0.0                  0.0   
9483                 0.0                 0.0                  1.0   
9484                 1.0                 0.0                  0.0   

      stage_diagnosis_IIB  stage_diagnosis_IIIA  stage_diagnosis_IIIB  \
0                     0.0                   0.0                   0.0   
1                     0.0                   0.0                   0.0   
2                     0.0                   0.0                   0.0   
3                     0.0                   1.0                   0.0   
4                     0.0                   0.0                   0.0   
...                   ...                   ...                   ...   
9480                  0.0                   0.0                   0.0   
9481                  0.0                   0.0                   0.0   
9482                  0.0                   0.0                   0.0   
9483                  0.0                   0.0                   0.0   
9484                  0.0                   0.0                   0.0   

      stage_diagnosis_IIIC  stage_diagnosis_IV  
0                      0.0                 0.0  
1                      0.0                 0.0  
2                      0.0                 0.0  
3                      0.0                 0.0  
4                      0.0                 0.0  
...                    ...                 ...  
9480                   1.0                 0.0  
9481                   0.0                 0.0  
9482

In [1274]:
tumor_patients_recaida["recaida"]

0       1.0
1       0.0
2       1.0
3       1.0
4       1.0
       ... 
9480    1.0
9481    1.0
9482    1.0
9483    1.0
9484    1.0
Name: recaida, Length: 9485, dtype: float64

In [1275]:
porcentaje_0 = (tumor_patients_recaida['recaida'] == 0).mean() * 100
porcentaje_1 = (tumor_patients_recaida['recaida'] == 1).mean() * 100

# Imprime los resultados
print(f'Porcentaje de m_category igual a 0: {porcentaje_0:.2f}%')
print(f'Porcentaje de m_category igual a 1: {porcentaje_1:.2f}%')

Porcentaje de m_category igual a 0: 12.03%
Porcentaje de m_category igual a 1: 87.97%


Verificamos que efectivamente no haya valores nulos:

In [1276]:
tumor_patients_recaida.isnull().sum()

ehr                     0
n_category              0
m_category              0
grade                   0
ductal                  0
lobular                 0
neoadjuvant             0
age                     0
diagnosis_age           0
dead                    0
er                      0
her2                    0
ki67                    0
pr                      0
recaida                 0
t_category_0            0
t_category_1            0
t_category_2            0
t_category_3            0
t_category_4            0
t_category_IS           0
stage_diagnosis_0       0
stage_diagnosis_IA      0
stage_diagnosis_IB      0
stage_diagnosis_IIA     0
stage_diagnosis_IIB     0
stage_diagnosis_IIIA    0
stage_diagnosis_IIIB    0
stage_diagnosis_IIIC    0
stage_diagnosis_IV      0
dtype: int64

Finalmente, creamos un daset normalizado para su uso en alguno de los modelos que usaremos. Habra otros modelos en los que no sea necesdaria la normalizacion como Decision Trees, Random Forest o Gradient Boosting en general. 

In [1277]:
from sklearn.preprocessing import MinMaxScaler

#Eliminamos el id del paciente
tumor_patients_recaida.drop(columns='ehr', inplace=True)

scaler = MinMaxScaler()
tumor_patients_recaida_norm = tumor_patients_recaida.copy()

#Automatizamos el proceso de identificar variables binarias y no tenerlas en cuenta para la normalizacion.
norm_cols = [col for col in tumor_patients_recaida_norm.columns if sorted(tumor_patients_recaida_norm[col].unique()) != [0, 1]]

# No queremos normalizar las columnas binarias.
tumor_patients_recaida_norm[norm_cols] = scaler.fit_transform(tumor_patients_recaida_norm[norm_cols])


In [1278]:
tumor_patients_recaida_norm

n_category  m_category  grade  ductal  lobular  neoadjuvant       age  \
0       0.000000         0.0    0.0     0.0      0.0          0.0  0.527027   
1       0.000000         0.0    0.5     1.0      0.0          0.0  0.608108   
2       0.000000         0.0    0.0     1.0      0.0          0.0  0.432432   
3       0.333333         0.0    1.0     0.0      1.0          1.0  0.608108   
4       0.333333         0.0    0.5     1.0      0.0          1.0  0.527027   
...          ...         ...    ...     ...      ...          ...       ...   
9480    1.000000         0.0    0.5     1.0      0.0          1.0  0.527027   
9481    0.333333         0.0    1.0     1.0      0.0          1.0  0.270270   
9482    0.000000         0.0    0.5     0.0      0.0          0.0  0.662162   
9483    0.000000         0.0    0.5     1.0      0.0          1.0  0.797297   
9484    0.000000         0.0    1.0     0.0      0.0          1.0  0.486486   

      diagnosis_age  dead   er  ...  t_category_IS  stage_diagnosis_0  \
0          0.506024   0.0  0.0  ...            1.0                1.0   
1          0.602410   0.0  1.0  ...            0.0                0.0   
2          0.469880   0.0  1.0  ...            0.0                0.0   
3          0.614458   0.0  1.0  ...            0.0                0.0   
4          0.542169   0.0  0.0  ...            0.0                0.0   
...             ...   ...  ...  ...            ...                ...   
9480       0.578313   0.0  1.0  ...            0.0                0.0   
9481       0.313253   0.0  1.0  ...            0.0                0.0   
9482       0.650602   0.0  1.0  ...            0.0                0.0   
9483       0.783133   0.0  0.0  ...            0.0                0.0   
9484       0.493976   0.0  1.0  ...            0.0                0.0   

      stage_diagnosis_IA  stage_diagnosis_IB  stage_diagnosis_IIA  \
0                    0.0                 0.0                  0.0   
1                    1.0                 0.0                  0.0   
2                    1.0                 0.0                  0.0   
3                    0.0                 0.0                  0.0   
4                    0.0                 1.0                  0.0   
...                  ...                 ...                  ...   
9480                 0.0                 0.0                  0.0   
9481                 0.0                 1.0                  0.0   
9482                 1.0                 0.0                  0.0   
9483                 0.0                 0.0                  1.0   
9484                 1.0                 0.0                  0.0   

      stage_diagnosis_IIB  stage_diagnosis_IIIA  stage_diagnosis_IIIB  \
0                     0.0                   0.0                   0.0   
1                     0.0                   0.0                   0.0   
2                     0.0                   0.0                   0.0   
3                     0.0                   1.0                   0.0   
4                     0.0                   0.0                   0.0   
...                   ...                   ...                   ...   
9480                  0.0                   0.0                   0.0   
9481                  0.0                   0.0                   0.0   
9482                  0.0                   0.0                   0.0   
9483                  0.0                   0.0                   0.0   
9484                  0.0                   0.0                   0.0   

      stage_diagnosis_IIIC  stage_diagnosis_IV  
0                      0.0                 0.0  
1                      0.0                 0.0  
2                      0.0                 0.0  
3                      0.0                 0.0  
4                      0.0                 0.0  
...                    ...                 ...  
9480                   1.0                 0.0  
9481                   0.0                 0.0  
9482                   0.0       

In [1279]:
tumor_patients_recaida.to_csv('recaida.csv')
tumor_patients_recaida_norm.to_csv('recaida_norm.csv')